src/array.vala

```diff
@@ -123,24 +163,99 @@ public class Array<T>: Object
         return shape;
     }
 
-    public Array<T>
-    slice (ssize_t[] from, ssize_t[] to)
+    public Array
+    slice ([CCode (array_length = false)] ssize_t[] from, [CCode (array_length = false)] ssize_t[] to)
+    {
+        return new Array (scalar_type,
+                          scalar_size,
+                          _shape_for_slice (from, to),
+                          _strides,
+                          data.slice ((int) _offset_for_index (from), (int) _offset_for_index (to)));
+    }
+
+    private inline void
+    _append_from_index (StringBuilder sb, ssize_t[] index)
     {
-        return new Array<T> (scalar_size,
-                                      _shape_from_slice (from, to),
-                                      _strides,
-                                      data.slice ((int) _offset_for_index (from), (int) _offset_for_index (to)));
+        sb.append_c ('\n');
+
+        // vector style
+        if (index.length == ndim - 1) {
+            sb.append_c ('[');
+            for (var i = 0; i < shape[index.length]; i++) {
+                if (i > 0)
+                    sb.append_c (' ');
+                // index and print the scalar
+                ssize_t[] subindex = index;
+                subindex += i;
+                sb.append (get_value (subindex).strdup_contents ());
+                if (i < shape[index.length] - 1)
+                    sb.append_c ('\n');
+            }
+            sb.append_c (']');
+        }
+
+        // matrix style
+        else if (index.length == ndim - 2) {
+            sb.append_c ('[');
+            // last dim is printed vertically
```

Comment:
I am not so sure about having a transpose here is a good idea. It will create a lot of confusions if any dimensions are handled differently... There were some lessons from numpy.matrix -- in general nowadays people are encouraged not to it.

Can we make sure the last dimension, when printed is continuous? This makes sure the printed version agrees with the memory when the array is 'tight'. Also the code is shorter. 

We just have to make sure the contraction operator ('tensor dot') is always explicit about which two axis-s are being contracted.